File: totosport_project/backend/app/schemas/tabellone.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class TabelloneBase(BaseModel):
    """
    Campi condivisi del tabellone (pronostici del giocatore).
    Tutti opzionali → consente salvataggi parziali (bozze) e, per il PATCH di
    modifica, di inviare solo i campi che cambiano (`model_dump(exclude_unset=True)`).
    """

    # Serie A
    scudetto_team: Optional[str] = None
    scudetto_points_guess: Optional[int] = None
    relegated_a_1: Optional[str] = None
    relegated_a_2: Optional[str] = None
    relegated_a_3: Optional[str] = None
    top_scorer_a: Optional[str] = None
    top_scorer_a_goals: Optional[int] = None
    # Serie B
    promoted_b_direct_1: Optional[str] = None
    promoted_b_direct_2: Optional[str] = None
    promoted_b_first_points: Optional[int] = None
    playoff_b_1: Optional[str] = None
    playoff_b_2: Optional[str] = None
    playoff_b_3: Optional[str] = None
    playoff_b_4: Optional[str] = None
    playoff_b_5: Optional[str] = None
    playoff_b_6: Optional[str] = None
    promoted_b_playoff: Optional[str] = None
    relegated_b_c_direct_1: Optional[str] = None
    relegated_b_c_direct_2: Optional[str] = None
    relegated_b_c_direct_3: Optional[str] = None
    playout_b_1: Optional[str] = None
    playout_b_2: Optional[str] = None
    relegated_b_c_playout: Optional[str] = None
    top_scorer_b: Optional[str] = None
    top_scorer_b_goals: Optional[int] = None
    # Coppe
    coppa_italia_winner: Optional[str] = None
    champions_winner: Optional[str] = None
    europa_winner: Optional[str] = None
    conference_winner: Optional[str] = None


class TablePredictionCreate(TabelloneBase):
    """POST /tabellone — compila o sovrascrive il tabellone."""


class TablePredictionModify(TabelloneBase):
    """PATCH /tabellone/me — solo i campi inviati vengono modificati."""


class TablePredictionOut(TabelloneBase):
    id: uuid.UUID
    player_id: uuid.UUID
    season_id: uuid.UUID
    mercato_penalty: int
    total_points: Optional[int] = None
    points_breakdown: Optional[dict] = None
    scored_at: Optional[datetime] = None
    submitted_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    # Calcolato dal router in base allo stato stagione (True se in 'mercato')
    is_modifiable: bool = False

    model_config = {"from_attributes": True}


# ─── Season Outcome (risultati reali, lato admin) ────────────────────────────────


class SeasonOutcomeBase(BaseModel):
    # Serie A
    scudetto_team: Optional[str] = None
    scudetto_points: Optional[int] = None
    relegated_a_1: Optional[str] = None
    relegated_a_2: Optional[str] = None
    relegated_a_3: Optional[str] = None
    top_scorer_a: Optional[str] = None
    top_scorer_a_goals: Optional[int] = None
    # Serie B
    promoted_b_direct_1: Optional[str] = None
    promoted_b_direct_2: Optional[str] = None
    promoted_b_first_points: Optional[int] = None
    playoff_b_1: Optional[str] = None
    playoff_b_2: Optional[str] = None
    playoff_b_3: Optional[str] = None
    playoff_b_4: Optional[str] = None
    playoff_b_5: Optional[str] = None
    playoff_b_6: Optional[str] = None
    promoted_b_playoff: Optional[str] = None
    playoffs_held: Optional[bool] = None
    relegated_b_c_direct_1: Optional[str] = None
    relegated_b_c_direct_2: Optional[str] = None
    relegated_b_c_direct_3: Optional[str] = None
    playout_b_1: Optional[str] = None
    playout_b_2: Optional[str] = None
    relegated_b_c_playout: Optional[str] = None
    playout_held: Optional[bool] = None
    top_scorer_b: Optional[str] = None
    top_scorer_b_goals: Optional[int] = None
    # Coppe
    coppa_italia_winner: Optional[str] = None
    champions_winner: Optional[str] = None
    europa_winner: Optional[str] = None
    conference_winner: Optional[str] = None


class SeasonOutcomeCreate(SeasonOutcomeBase):
    """POST /admin/season-outcome — inserisce/aggiorna i risultati reali."""


class SeasonOutcomeOut(SeasonOutcomeBase):
    id: uuid.UUID
    season_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: totosport_project/frontend/src/types/tabellone.ts
// Voci del tabellone annuale (REGOLAMENTO §2). Tutti i campi sono opzionali
// per consentire salvataggi parziali (bozze) e PATCH parziali.
export interface TabelloneFields {
  // Serie A
  scudetto_team: string | null
  scudetto_points_guess: number | null
  relegated_a_1: string | null
  relegated_a_2: string | null
  relegated_a_3: string | null
  top_scorer_a: string | null
  top_scorer_a_goals: number | null
  // Serie B
  promoted_b_direct_1: string | null
  promoted_b_direct_2: string | null
  promoted_b_first_points: number | null
  playoff_b_1: string | null
  playoff_b_2: string | null
  playoff_b_3: string | null
  playoff_b_4: string | null
  playoff_b_5: string | null
  playoff_b_6: string | null
  promoted_b_playoff: string | null
  relegated_b_c_direct_1: string | null
  relegated_b_c_direct_2: string | null
  relegated_b_c_direct_3: string | null
  playout_b_1: string | null
  playout_b_2: string | null
  relegated_b_c_playout: string | null
  top_scorer_b: string | null
  top_scorer_b_goals: number | null
  // Coppe
  coppa_italia_winner: string | null
  champions_winner: string | null
  europa_winner: string | null
  conference_winner: string | null
}

export interface TablePredictionOut extends TabelloneFields {
  id: string
  player_id: string
  season_id: string
  mercato_penalty: number
  total_points: number | null
  points_breakdown: Record<string, number> | null
  scored_at: string | null
  submitted_at: string | null
  created_at: string
  updated_at: string
  is_modifiable: boolean
}

export type TabelloneInput = Partial<TabelloneFields>

// Chiavi dei campi nell'ordine di compilazione, raggruppate per sezione.
export type TabelloneFieldKey = keyof TabelloneFields
